### ISSUE TITLE
Support for nested properties

### ISSUE THREAD

**Issue by hekonsek** [mapstruct/mapstruct#65]
It would be nice if MapStruct could support compound properties mappings. For example I would like to map `Person.Address.street` property to `PersonDto.addressStreet`.

First of all `@Mapping` annotation could support dot-notation of properties (for example `@Mapping(source="address.street", target="addressStreet"))`. However ideally MapStruct could automagically try to map `Person.Address.street` to `PersonDto.addressStreet`, if there is no `Person.addressStreet` property in the source bean and no `PersonDto.Address.street` in the target bean.

**Comment by devj87** [comment]
Hi same issue for me I need to map as source a property from an object within in a List.

Having:

I'm using MapStruct to create my DTOs, so I have a problem to set target attribute from an Object within collection:

My DTO is Employee which hava a department attribute.

The department name value is in DepartmentLanguage class which is a collection within a Department entity class has departmentLanguage attribute

So how can I set department target field over this class design?

Following my classes:

EmployeeTo
-------------------------------------------
		public class EmployeeTo implements Serializable {
			
		  private String department;
		  
		 }
			




Department
-----------------------
    @Entity
    @Table(name="departments")
    @NamedQuery(name="Department.findAll", query="SELECT d FROM Department d")
    public class Department implements Serializable {
    	private static final long serialVersionUID = 1L;
    
    
    	//bi-directional many-to-one association to DepartmentsLanguage
    	@OneToMany(mappedBy="department")
    	private List<DepartmentsLanguage> departmentsLanguages;
    
    
    	public List<DepartmentsLanguage> getDepartmentsLanguages() {
    		return this.departmentsLanguages;
    	}
    
    	public void setDepartmentsLanguages(List<DepartmentsLanguage> departmentsLanguages) {
    		this.departmentsLanguages = departmentsLanguages;
    	}   
    
    }

DepartmentLanguage
-------------------------------------
		@Table(name="departments_languages")
		@NamedQuery(name="DepartmentsLanguage.findAll", query="SELECT d FROM DepartmentsLanguage d")
		public class DepartmentsLanguage implements Serializable {
			private static final long serialVersionUID = 1L;

			@EmbeddedId
			private DepartmentsLanguagePK id;

			@Column(nullable=false, length=255)
			private String description;

			@Column(nullable=false, length=100)
			private String name;

			//bi-directional many-to-one association to Department
			@ManyToOne(fetch=FetchType.EAGER)
			@JoinColumn(name="department_id", nullable=false, insertable=false, updatable=false)
			private Department department;

			//bi-directional many-to-one association to Language
			@ManyToOne(fetch=FetchType.EAGER)
			@JoinColumn(name="language_id", nullable=false, insertable=false, updatable=false)
			private Language language;

			public DepartmentsLanguage() {
			}
		}

Thanks a lot!

**Comment by anchitanc** [comment]
Hi 
I am also looking for both source and Target nested mapping.

```
@Mapper(componentModel = "spring", uses = {})
   public interface CategoriesMapper {
      @Mapping(source = "regionsid.id", target = "regionsid.id")
      CategoriesDTO categoriesToCategoriesDTO(Categories categories);  
   } 

    public class Categories implements Serializable {
        private Regions regionsid;
        public Regions getRegionsid() {
            return regionsid;
        }

        public void setRegionsid(Regions regions) {
            this.regionsid = regions;
        }

    }

    public class CategoriesDTO implements Serializable {
        private RegionsDTO regionsid;
        public RegionsDTO getRegionsid() {
            return regionsid;
        }

        public void setRegionsid(RegionsDTO regions) {
            this.regionsid = regions;
        }

    }

    public class Regions implements Serializable {
        private Long id;
    }

    public class RegionsDTO implements Serializable {
        private Long id;
    } 
```

i am getting below error -
`CategoriesMapper.java:28: error: Unknown property "regionsid.id" in return type.`

I have angularJS view which required the data in nested properties. I can't send domain class because it has so many unnecessary information.

**Comment by agudian** [comment]
@anchitanc: either try the latest version 1.1.0-CR2 (Final is coming up this weekend, probably), or just define another mapping method that maps from Regions to RegionsDto (and remove the `@Mapping` annotation).

**Comment by devj87** [comment]
thanks for reply @anchitanc my issue is to access an attribute from a Collection within a class
e.g
```
private List<DepartmentsLanguage> departmentsLanguages;
```

so I need to get the 1st index from it and access to "name" attribute:, I tried both with

```
@Mapping(source="departmentsLanguages[0].name", target="department"))
```
and
```
@Mapping(expression="java(departmentsLanguages.get(0).getName())", target="department"))
```

But it doesn't work

**Comment by anchitanc** [comment]
@agudian  i have tried latest 1.1.0-CR2, its seem error is resolved but i am stuck with another error like below, what if i have Set<RegionsDTO>

Source -
`private Set<Regions> regions = new HashSet<>();`

Target -
 `private Set<RegionsDTO> regions = new HashSet<>();`

`CategoriesMapper.java:36: error: Can't map property "java.util.Set<com.equidity.xboard.service.dto.RegionsDTO> regions" to "java.util.Set<com.equidity.xboard.domain.Regions> regions". Consider to declare/implement a mapping method: "java.util.Set<com.equidity.xboard.domain.Regions> map(java.util.Set<com.equidity.xboard.service.dto.RegionsDTO> value)".``

**Comment by agudian** [comment]
@devj87 just check the generated code of your last example and you'll see that the expression within `java( xx )` is used as you wrote it. Add compilable code there and it should work.

@anchitanc: just add a mapping method to map from RegionsDto to Regions (check our documentation on why it's enough to add that method for mapping sets of those types 😉).

@devj87 and @anchitanc, I'd appreciate it if you could use our mailing list for further questions, as I don't think that these questions have that much to do with the implemented feature described in this issue.

**Comment by devj87** [comment]
Thanks @agudian solved using expression in 1.1.0-CR2 version
Next time I'll ask question in separated area, can you post me link please?

**Comment by sjaakd** [comment]
https://groups.google.com/forum/m/?fromgroups#!forum/mapstruct-users
